Bluetooth: Log: Fix Bluetooth string log functions with LOG_IMMEDIATE

joerchan opened the pull request:
Fix Bluetooth string log functions with LOG_IMMEDIATE enabled if the
same log function was called twice in a print statement this would
overwrite the temporary buffer.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

Vudentz commented on subsys/bluetooth/common/log.c:
```diff
@@ -40,18 +47,32 @@ const char *bt_hex_real(const void *buf, size_t len)
 
 const char *bt_addr_str_real(const bt_addr_t *addr)
 {
+#ifdef CONFIG_LOG_IMMEDIATE
```
Is this a problem only for LOG_IMMEDIATE? There could be valid reasons to have multiple calls to these functions i.e. generate a string for settings, etc.

joerchan replied:
Yes, it is only a problem for LOG_IMMEDIATE, otherwise log_strdup will create a copy.

Vudentz replied:
You are saying log_strdup don't work when LOG_IMMEDIATE is set? Isn't it better to fix that, otherwise we have to duplicate this sort of buffering everywhere.

joerchan replied:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/logging/log_core.c#L846
If not log immediate the same pointer is returned.

I think it is only a problem for these bt_*_str_real functions though, so I don't think we have to duplicate this sort of buffering anywhere else. This is how these functions was implement before the logging and log_strdup was introduced.

nordic-krch replied:
I wonder if the way this buffer is used in bluetooth is not causing issues. Considering example:
```
LOG_INF("%s connected to %s",
         log_strdup(bt_addr_str_real(local_addr)),
         log_strdup(bt_addr_str_real(remote_addr)));
```
`LOG_IMMEDIATE` and proposed `LOG_MINIMAL` are turning log messages into simple, synchronous printf like calls and this approach is not ready for that.

joerchan replied:
Ok, I'll accept that explanation and update the host to not use log_strdup in this way anymore.

Vudentz commented on subsys/bluetooth/common/log.c:
```diff
@@ -40,18 +47,32 @@ const char *bt_hex_real(const void *buf, size_t len)
+	static u8_t cur = 0;
+	static char bufs[4][BT_ADDR_LE_STR_LEN];
+	char *str = bufs[cur];
+	cur = (cur + 1) % 4;
```
I wonder if we tap into the logging buffer to do such thing, though we don't free the memory anywhere but this is generating duplicated code as well.

joerchan replied:
I guess you mean the log_strdup here? In which case log_strdup will return the same pointer under LOG_IMMEDIATE. Which is what is causing invalid (and confusing) logging output.

Vudentz replied:
The idea was to keep the log_strdup working even with LOG_IMMEDIATE that way we don't have to maintain 2 different code paths, though it appears we would be introducing a third way to log: https://github.com/zephyrproject-rtos/zephyr/pull/19322

joerchan replied:
We could take this approach and just move it to the log_strdup function when using LOG_IMMEDIATE, instead of just returning the string like it does now.

joerchan replied:
@Vudentz With minimal logging maybe we could get rid of LOG_IMMEDIATE, it seems to be problematic in many ways.

Vudentz replied:
Indeed, Id hope we do that so perhaps it is good idea to drop a comment in the PR.

joerchan replied:
Already did. Let's hope they go for it.

Vudentz replied:
I guess one of the problem is the support of backends, Im not sure if the intention was not disable the backends as well but lets see their response.

nordic-krch replied:
@Vudentz see my comment: https://github.com/zephyrproject-rtos/zephyr/pull/19322#issuecomment-534965754 imo, `LOG_MINIMAL` is not fixing issues with `LOG_IMMADIATE` but rather cuts features and reduces memory footprint.

Vudentz replied:
Well regardless what LOG_MINIMAL does all we need to do is to make log_strdup always dupping the string with is own buffers, that is fairly minimal and would work for any code using log_strdup with or without LOG_MINIMAL or LOG_IMMEDIATE.

nordic-krch replied:
do i understand correctly, you meant that string duplicating feature should be present in minimal/immediate mode?

Vudentz replied:
Yes, absolutely Im not sure why it was thought not to be the case since it is quite obvious one could pass a local variable to log_strdup and then suddenly it doesn't work depending on the config options.

carlescufi commented:
@nordic-krch we'd really appreciate your input on this PR.

joerchan committed:
Bluetooth: Host: Fix usage of re-using Bluetooth log buffers

Fix calling bt_hex and bt_addr_le_str multiple times in the same logging
call could result in string overwritten since log_strdup is not
guaranteed to duplicate the string buffer in all logging configurations.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

Vudentz reviewed (approved):
LGTM
